tutorials_nccl/poststack_nccl.py

```diff
@@ -0,0 +1,202 @@
+r"""
+Post Stack Inversion - 3D with NCCL
+============================================
+This tutorial is an extension of the :ref:`sphx_glr_tutorials_poststack.py` tutorial where PyLops-MPI is run in multi-GPU setting with GPUs communicating via NCCL.
+"""
+
+import numpy as np
+import cupy as cp
+from scipy.signal import filtfilt
+from matplotlib import pyplot as plt
+from mpi4py import MPI
+
+from pylops.utils.wavelets import ricker
+from pylops.basicoperators import Transpose
+from pylops.avo.poststack import PoststackLinearModelling
+
+import pylops_mpi
+import pylops_mpi.utils
+import pylops_mpi.utils._nccl
+
+###############################################################################
+# NCCL communication can be easily initialized with
+# :py:func:`pylops_mpi.utils._nccl.initialize_nccl_comm` operator.
+# One can think of this as GPU-counterpart of :code:`MPI.COMM_WORLD`
+
+plt.close("all")
+nccl_comm = pylops_mpi.utils._nccl.initialize_nccl_comm()
+rank = MPI.COMM_WORLD.Get_rank()
+
+###############################################################################
+# Let's start by defining all the parameters required by the
+# :py:func:`pylops.avo.poststack.PoststackLinearModelling` operator.
+# Note that this section is exactly the same as the one in the MPI example as we will keep using MPI for transfering metadata (i.e., shapes, dims, etc.)
+
+# Model
+model = np.load("../testdata/avo/poststack_model.npz")
+x, z, m = model['x'][::3], model['z'], np.log(model['model'])[:, ::3]
+
+# Making m a 3D model
+ny_i = 20  # size of model in y direction for rank i
+y = np.arange(ny_i)
+m3d_i = np.tile(m[:, :, np.newaxis], (1, 1, ny_i)).transpose((2, 1, 0))
+ny_i, nx, nz = m3d_i.shape
+
+# Size of y at all ranks
+ny = MPI.COMM_WORLD.allreduce(ny_i)
+
+# Smooth model
+nsmoothy, nsmoothx, nsmoothz = 5, 30, 20
+mback3d_i = filtfilt(np.ones(nsmoothy) / float(nsmoothy), 1, m3d_i, axis=0)
+mback3d_i = filtfilt(np.ones(nsmoothx) / float(nsmoothx), 1, mback3d_i, axis=1)
+mback3d_i = filtfilt(np.ones(nsmoothz) / float(nsmoothz), 1, mback3d_i, axis=2)
+
+# Wavelet
+dt = 0.004
+t0 = np.arange(nz) * dt
+ntwav = 41
+wav = ricker(t0[:ntwav // 2 + 1], 15)[0]
```

Comment:
You should add `wav =  cp.asarray(wav)` to make sure it is on the GPU... it still works as internally in PyLops Convolve operator we check if the input is on GPU and filter is not, and we move it...